packages/panzer/mini-em/example/BlockPrec/main.cpp

```diff
@@ -0,0 +1,691 @@
+#include "MiniEM_BCStrategy_Factory.hpp"
+#include "MiniEM_ClosureModel_Factory_TemplateBuilder.hpp"
+#include "MiniEM_AddFieldsToMesh.hpp"
+#include "MiniEM_OperatorRequestCallback.hpp"
+#include "MiniEM_FullMaxwellPreconditionerFactory.hpp"
+
+#include <string>
+#include <iostream>
+
+Teuchos::RCP<Teuchos::ParameterList> maxwellParameterList(const int basis_order);
+std::vector<panzer::BC> homogeneousBoundaries();
+std::vector<panzer::BC> auxiliaryBoundaries();
+Teuchos::RCP<Teuchos::ParameterList> auxOpsParameterList(const int basis_order);
+Teuchos::RCP<Teuchos::ParameterList> maxwellSolverParameterList(const bool use_ilu, const bool use_refmaxwell, const bool print_diagnostics);
+void createExodusFile(const std::vector<Teuchos::RCP<panzer::PhysicsBlock> >& physicsBlocks,
+                      Teuchos::RCP<panzer_stk::STK_MeshFactory> mesh_factory,
+                      Teuchos::RCP<panzer_stk::STK_Interface> mesh,
+                      const bool & exodus_out);
+Teuchos::RCP<panzer::ResponseLibrary<panzer::Traits> >
+buildSTKIOResponseLibrary(const std::vector<Teuchos::RCP<panzer::PhysicsBlock> > & physicsBlocks,
+    const Teuchos::RCP<panzer::LinearObjFactory<panzer::Traits> > & linObjFactory,
+    const Teuchos::RCP<panzer::WorksetContainer> & wkstContainer,
+    const Teuchos::RCP<panzer::UniqueGlobalIndexerBase> & globalIndexer,
+    const panzer::ClosureModelFactory_TemplateManager<panzer::Traits> & cm_factory,
+    const Teuchos::RCP<panzer_stk::STK_Interface> & mesh,
+    const Teuchos::ParameterList & closure_model_pl);
+void writeToExodus(double time_stamp,
+    const Teuchos::RCP<const Thyra::VectorBase<double> > & x,
+    const panzer::ModelEvaluator<double> & model,
+    panzer::ResponseLibrary<panzer::Traits> & stkIOResponseLibrary,
+    panzer_stk::STK_Interface & mesh);
+
+/********************************************************************************
+ * Sets up an electromagetics problem driven by a simple Gaussian current pulse
+ * on the domain [0,1]^3. First order Maxwell equations with edge-face
+ * discretization for E,B. Backward Euler time-stepping with fixed CFL. Linear
+ * systems solved with Belos GMRES using augmentation based block preconditioner
+ * through Teko with multigrid subsolves from MueLu.
+ *
+ * This is meant to test the components of the Tpetra linear solver stack 
+ * required by EMPIRE-EM
+ *
+ * Command-line arguments:
+ *   x-elements, y-elements,z-elements: # elements in each direction
+ *   basis-order: order of finite element bases
+ *   cfl: CFL with respect to speed of light. CFL = c*dt/min{dx,dy,dz}
+ *   workset-size: size of worksets
+ *   use-ilu: use ILU as smoother for E-field Schur complement
+ *
+ * ******************************************************************************
+ */
+
+int main(int argc,char * argv[])
+{
+
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  using Teuchos::rcp_dynamic_cast;
+
+  Kokkos::initialize(argc, argv);
```

Comment:
Please initialize Kokkos _after_ initializing MPI.  This ensures that MPI correctly binds processes to sockets / cores / etc., before Kokkos attempts to bind threads.